### ISSUE TITLE
Add support for loading plugins as shared libraries

### ISSUE THREAD

**Issue by tsandall** [open-policy-agent/opa#723]
[Golang supports building shared libraries now.](https://golang.org/pkg/plugin/) We can extend the runtime to load plugins dynamically. This would allow people to extend OPA without recompiling it.

**Comment by tsandall** [comment]
This looks good to me.

I just realized there's one gotcha in the runtime initialization.

[Currently the plugins are loaded after Rego files supplied to the runtime are compiled.](https://github.com/open-policy-agent/opa/blob/master/runtime/runtime.go#L151
) As a result, if the these plugins are registering built-ins, the compile step will fail. The manager also maintains an instance of the compiler that's shared across all of the plugins.

The existing plugin support was added to help with management functionality (e.g., bundle downloading, status reporting, etc.) It wasn't intended to be used for registering new built-ins.

We'll either need to refactor the current plugin implementation to handle this case, or, distinguish between built-in functions and plugins.

**Comment by vrnmthr** [comment]
I was planning on distinguishing between builtin functions and plugins

**Comment by vrnmthr** [comment]
Builtin functions could be registered as such:

```go
package main

Builtin := &ast.Builtin {
    Name: ...
    Decl: ...
}

Function := func(a, b ast.Value) (ast.Value, error) {
    // do whatever here
}
```

On our end, roughly: 
```go
func LoadBuiltin(path string) (error) {
    mod, err := plugin.Load(path)
    builtinSym, err := mod.Lookup("Builtin")
    functionSym, err := mod.Lookup("Function")
    if builtin == nil || function == nil {
        return err
    }
    builtin, err := builtinSym.(*ast.Builtin)
    switch function := functionSym.(type) {
    case topdown.FunctionalBuiltin1:
        topdown.RegisterFunctionalBuiltin1(builtin.Name, function)
    case topdown.FunctionalBuiltin2:
        topdown.RegisterFunctionalBuiltin2(builtin.Name, function)
    case topdown.FunctionalBuiltin3:
        topdown.RegisterFunctionalBuiltin3(builtin.Name, function)
    }
}
```

**Comment by tsandall** [comment]
@vrnmthr is working on this. The initial PR is up #829.

**Comment by tsandall** [comment]
Merged in #883.
